src/snowflake/connector/nanoarrow_cpp/ArrowIterator/nanoarrow_arrow_iterator.pyx

```diff
@@ -0,0 +1,255 @@
+            int64_t arrow_bytes_size,
+            PyObject* use_numpy
+        ) except +
+
+cdef extern from "CArrowTableIterator.hpp" namespace "sf":
+    cdef cppclass CArrowTableIterator(CArrowIterator):
+        CArrowTableIterator(
+            PyObject* context,
+            char* arrow_bytes,
+            int64_t arrow_bytes_size,
+            bint number_to_decimal,
+        ) except +
+
+
+cdef class EmptyPyArrowIterator:
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+       raise StopIteration
+
+    def init(self, str iter_unit, bint number_to_decimal):
+        pass
+
+
+cdef class PyArrowIterator(EmptyPyArrowIterator):
+    cdef object context
+    cdef CArrowIterator* cIterator
+    cdef str unit
+    cdef shared_ptr[ReturnVal] cret
+    cdef object use_dict_result
+    cdef object cursor
+    cdef vector[uintptr_t] nanoarrow_Table
+    cdef vector[uintptr_t] nanoarrow_Schema
+    cdef object table_returned
+    cdef char* arrow_bytes
+    cdef int64_t arrow_bytes_size
+
+    # this is the flag indicating whether fetch data as numpy datatypes or not. The flag
+    # is passed from the constructor of SnowflakeConnection class. Note, only FIXED, REAL
+    # and TIMESTAMP_NTZ will be converted into numpy data types, all other sql types will
+    # still be converted into native python types.
+    # https://docs.snowflake.com/en/user-guide/sqlalchemy.html#numpy-data-type-support
+    cdef object use_numpy
+    cdef object number_to_decimal
+    cdef object pyarrow_table
+
+    def __cinit__(
+            self,
+            object cursor,
+            object arrow_bytes,
+            object arrow_context,
+            object use_dict_result,
+            object numpy,
+            object number_to_decimal,
+    ):
+        self.context = arrow_context
+        self.cIterator = NULL
+        self.use_dict_result = use_dict_result
+        self.cursor = cursor
+        self.use_numpy = numpy
+        self.number_to_decimal = number_to_decimal
+        self.pyarrow_table = None
+        self.table_returned = False
+        self.arrow_bytes = <char*>arrow_bytes
+        self.arrow_bytes_size = len(arrow_bytes)
+
+    def __dealloc__(self):
+        del self.cIterator
+
+    def __iter__(self):
+        return self
+
+
+cdef class PyArrowRowIterator(PyArrowIterator):
+    def __cinit__(
+        self,
+        object cursor,
+        object py_inputstream,
+        object arrow_context,
+        object use_dict_result,
+        object numpy,
+        object number_to_decimal,
+    ):
+        super().__init__(cursor, py_inputstream, arrow_context, use_dict_result, numpy, number_to_decimal)
+        if self.cIterator is not NULL:
+            return
+
+        self.cIterator = new CArrowChunkIterator(
+            <PyObject *> self.context,
+            self.arrow_bytes,
+            self.arrow_bytes_size,
+            <PyObject *> self.use_numpy
+        ) \
+            if not self.use_dict_result \
```

Comment:
nit: how about `new DictCArrowChunkIterator(...) if self.use_dict_result else new CArrowChunkIterator(...)`?

Reply:
I did this because in most scenarios CArrowChunkIterator is used so the code doesn't need to go to the `else` statement. I'm wondering moving this to `else` will slight perf overhead.

I prefer just keeping it here for now

Reply:
I'm not sure about performance, but I find it not natural to do a mental negation when both true/false condition has some actions.

Reply:
I can change it and do a perf test, if no significant change I'm glad to change it

Reply:
I can't imagine there will be a significant performance difference for moving the case arms around.  That said, in general, and all else being equal (enough caveats?) if there are actions for both arms (rather that say an early exit), I always prefer putting the positive case in the `if` arm and the negative case.  The principle being that it's much harder to reason about double negatives.